Assets/Scripts/StandardSamples/Services/AchievementsService.cs

```diff
@@ -348,34 +353,62 @@ public Texture2D GetAchievementLockedIconTexture(string achievementId)
         /// The icon texture associated with the given achievement ID and
         /// determined by the given URI selector.
         /// </returns>
-        private Texture2D GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
+        private async Task<Texture2D> GetAchievementIconTexture(string achievementId, Func<DefinitionV2, string> uriSelector)
         {
-            Texture2D iconTexture = null;
+            Texture2D textureFromBytes = null;
+            
             foreach (var achievementDef in _achievements)
             {
                 if (achievementDef.AchievementId != achievementId)
                     continue;
 
-                if (_downloadCache.TryGetValue(uriSelector(achievementDef), out byte[] iconBytes) && null != iconBytes)
+                var uri = uriSelector(achievementDef);
+                byte[] iconBytes = null;
+
+                // Download the data
+                if (!_downloadCache.ContainsKey(uri))
                 {
-                    Texture2D textureFromBytes = new(2, 2);
-                    if (textureFromBytes.LoadImage(iconBytes))
+                    TaskCompletionSource<byte[]> downloadTcs = new();
+
+                    GetAndCacheData(uri, data =>
                     {
-                        iconTexture = textureFromBytes;
-                    }
-                    else
+                        if (data.result == UnityWebRequest.Result.Success)
+                        {
+                            downloadTcs.SetResult(data.data);
```

Comment:
Does the task completion source end when it has its data set? Or will the below code block of logging a warning and then setting the result still run?

Reply:
SetResult does not end the task completion source.